### PR TITLE
Retrieve spectra over a time range 

### PR DESCRIPTION
**Pull request by scizen9**
Not throwing errors, but not obviously filtering correctly on the range.

1. modified db_demo.yaml to change dates for 'unsaved_copy' versions of spectra from 2019-10-24 to 2019-10-28
2. added SpectrumRangeHandler to app_server.py, __init__.py
3. implemented SpectrumRangeHandler in spectrum.py using PhotometryRangeHandler as template
4. changed mjd filter to utc filter
5. test with:` http://localhost:5000/api/spectrum/range?instrument_ids=[2]&min_date=2019-10-22T00:00:00&max_date=2019-10-26T00:00:00`
6. no errors, but returns all four spectra in the demo: i.e. _not_ filtering based on date

Recommended that I modify model code for Spectrum(Base) to include arrow type for dates.  Still learning what this is.

-scizen9

### CODE REVIEW

File: skyportal/handlers/api/spectrum.py [hunk 1]
```diff
@@ -25,11 +26,40 @@
...
+        'instrument_id': spec.instrument_id,
+        'instrument_name': spec.instrument.name,
+        'origin': spec.origin,
+        'id': spec.id,
+        'groups': spec.groups,
+        'reducers': spec.reducers,
+        'observers': spec.observers
+    }
+
+    return return_value
```
Comment (dannygoldstein) [inline skyportal/handlers/api/spectrum.py:60]:
```suggestion
```

File: skyportal/handlers/api/spectrum.py [hunk 2]
```diff
@@ -547,3 +577,43 @@ def get(self, obj_id):
...
+        if instrument_ids is not None:
+            query = query.filter(Spectrum.instrument_id.in_(instrument_ids))
+        if min_date is not None:
+            utc = Time(min_date, format='isot', scale='utc')
+            query = query.filter(Spectrum.observed_at >= utc)
+        if max_date is not None:
+            utc = Time(max_date, format='isot', scale='utc')
+            query = query.filter(Spectrum.observed_at <= utc)
+
+        output = [serialize(p) for p in query]
```
Comment (dannygoldstein) [inline skyportal/handlers/api/spectrum.py:616]:
```suggestion
```

File: skyportal/handlers/api/spectrum.py [hunk 3]
```diff
@@ -547,3 +577,43 @@ def get(self, obj_id):
...
+        return self.success(data=output)
```
Comment (dannygoldstein) [inline skyportal/handlers/api/spectrum.py:617]:
```suggestion
        return self.success(data=query.all())
```

File: skyportal/handlers/api/spectrum.py [hunk 4]
```diff
@@ -547,3 +577,43 @@ def get(self, obj_id):
...
+        )
+
+        if instrument_ids is not None:
+            query = query.filter(Spectrum.instrument_id.in_(instrument_ids))
+        if min_date is not None:
+            utc = Time(min_date, format='isot', scale='utc')
+            query = query.filter(Spectrum.observed_at >= utc)
+        if max_date is not None:
+            utc = Time(max_date, format='isot', scale='utc')
+            query = query.filter(Spectrum.observed_at <= utc)
```
Comment (dannygoldstein) [inline skyportal/handlers/api/spectrum.py:614]:
```suggestion
            query = query.filter(Spectrum.observed_at <= utc.isot)
```

File: skyportal/handlers/api/spectrum.py [hunk 5]
```diff
@@ -547,3 +577,43 @@ def get(self, obj_id):
...
+            .query(Spectrum)
+            .join(GroupSpectrum)
+            .filter(GroupSpectrum.group_id.in_(gids))
+        )
+
+        if instrument_ids is not None:
+            query = query.filter(Spectrum.instrument_id.in_(instrument_ids))
+        if min_date is not None:
+            utc = Time(min_date, format='isot', scale='utc')
+            query = query.filter(Spectrum.observed_at >= utc)
```
Comment (dannygoldstein) [inline skyportal/handlers/api/spectrum.py:611]: @scizen9  This should fix the bug you are seeing vv
```suggestion
            query = query.filter(Spectrum.observed_at >= utc.isot)
```

**Review by dannygoldstein — CHANGES_REQUESTED** [review]

File: skyportal/handlers/api/spectrum.py (hunk 3 above)
Comment (scizen9) [reply skyportal/handlers/api/spectrum.py:617]: Nice!  I am not familiar with query functions.  This certainly tightens up the code.

File: skyportal/handlers/api/spectrum.py (hunk 2 above)
Comment (scizen9) [reply skyportal/handlers/api/spectrum.py]: accepted

File: skyportal/handlers/api/spectrum.py (hunk 3 above)
Comment (scizen9) [reply skyportal/handlers/api/spectrum.py]: accepted

File: skyportal/handlers/api/spectrum.py (hunk 4 above)
Comment (scizen9) [reply skyportal/handlers/api/spectrum.py]: Tried these changes, but still does not seem to be filtering.

File: skyportal/handlers/api/spectrum.py (hunk 5 above)
Comment (scizen9) [reply skyportal/handlers/api/spectrum.py]: Tried these changes, but still does not seem to be filtering.

**Comment by scizen9** [comment]
I don't understand the conflict.  It's just my change.  I don't see any conflict.

**Comment by scizen9** [comment]
Fixed the conflict and merged the fix.

File: skyportal/handlers/api/spectrum.py [hunk 6]
```diff
@@ -587,3 +588,33 @@ def get(self, obj_id):
                 )
 
         return self.success(data={'obj_id': obj.id, 'spectra': return_values})
+
+
+class SpectrumRangeHandler(BaseHandler):
+    @auth_or_token
+    def get(self):
+        """Docstring appears below as an f-string."""
```
Comment (acrellin) [inline skyportal/handlers/api/spectrum.py:596]: Still needs docstring, preferably inline right here as opposed to adding it below

**Review by acrellin — APPROVED** [review]
Needs a docstring, otherwise looks good!

File: skyportal/tests/api/test_spectrum.py [hunk 7]
```diff
@@ -6,6 +6,74 @@
 import datetime
 
 
+def test_token_user_get_range_spectrum(
```
Comment (dannygoldstein) [inline skyportal/tests/api/test_spectrum.py:9]: test improperly formatted date string 

test that the following ISO date strings for `min_date` / `max_date` all work


2020-12-1
2020-12-18T03:52:47+00:00
2020-12-18T03:52:47Z
20201218T035247Z

Comment (scizen9) [reply skyportal/tests/api/test_spectrum.py:9]: That last example throws an error for astropy.time.Time with format='isot'.  Tests for the others were implemented.

File: skyportal/handlers/api/spectrum.py (hunk 6 above)
Comment (scizen9) [reply skyportal/handlers/api/spectrum.py]: I've added a docstring, but please check the format.

File: skyportal/tests/api/test_spectrum.py (hunk 7 above)
Comment (dannygoldstein) [reply skyportal/tests/api/test_spectrum.py:9]: How about 

2020-12-18T03:52:47+07:00

?

Does it handle the time zone correctly?

Comment (scizen9) [reply skyportal/tests/api/test_spectrum.py:9]: I also added a test for when the parameter instrument_ids is not present.

Comment (scizen9) [reply skyportal/tests/api/test_spectrum.py:9]: Yes, but you have to use the html code '&plus;' and not the character '+', which generates a space.

**Review by dannygoldstein — APPROVED** [review]

**Pull request merged**